### PR TITLE
VACMS-14205 Fix profile required fields. #14026

### LINKED ISSUE #14205: Editors unable to save changes to Staff Profile pages
## Describe the defect
Helpdesk has received three separate reports of Staff Profiles not saving changes when 'published' . I have seen this myself on VA Sioux Falls and VA Jackson health care in Production and in Staging - In the Recent Changes list, my changes do not appear -- the third report comes from VA Greater Los Angeles health care.

## To Reproduce
1. Go to 'https://prod.cms.va.gov/va-jackson-health-care' and log in as log in as Camille.Ballentine@va.gov
2. Navigate to Staff Profiles ([Jason C. Cain i](https://staging.cms.va.gov/jackson-health-care/staff-profiles/jason-c-cain)s the prime example) 
3. Make a change, then Publish
4. See error - The change is not saved, and the edit does not appear in the 'Recent Changes' list.

## AC / Expected behavior
When a change is made to a Staff profile by an editor, that change should be saved upon Publish and/or Draft

## Screenshots
Camille has removed the 'Acting' from the title and the Phone number several times but the changes do not 'stick'
image

 ## Additional context
Other reports are from 

Amber.Bailey-Froke@va.gov at VA Sioux Falls health care
Lauren.Bolanos@va.gov at VA Greater Los Angeles health care

## Labels
(You can delete this section once it's complete)
- [x] Issue type (red) (defaults to "Defect")
- [ ] CMS subsystem (green)
- [ ] CMS practice area (blue)
- [x] CMS workstream (orange) (not needed for bug tickets)
- [ ] CMS-supported product (black)

### Team
Please check the team(s) that will do this work.

- [x] `CMS Team`
- [ ] `Public Websites`
- [ ] `Facilities`
- [ ] `User support`

### PR DESCRIPTION
**Pull request by swirtSJW**
## Description

Closes #14205 

relates to #12187

## QA steps

What needs to be checked to prove this works?  
What needs to be checked to prove it didn't break any related things?  
What variations of circumstances (users, actions, values) need to be checked?  

As content editor edgar.cruse 
1. visit https://pr14233-hghtp6e3u8qmoo00kebzk48sppmv1isn.ci.cms.va.gov/tuscaloosa-health-care/staff-profiles/james-hardemon
2. Edit the profile
3. Check the box for "Create profile page with biography"
4. delete any values in the First sentence and Body text
5. **UNCheck** the box for "Create profile page with biography"
6. Save the page with a revision log message
   - [x] Validate that it saves without error.
7. Edit the node.
8. Check the box for "Create profile page with biography"
9. Save the page with a revision log message
   - [x] Validate save is prevented with a notificationpointing to "First sentence" field (HTML 5 validation happens before save even occurs)
10. Add content to the "First sentence" field
11. Save the page with a revision log message
   - [x] Validate save is prevented with a notification pointing to "First sentence" field  (This is prevented after a true save attempt, not HTML5 pre-save)  **Note:** The jump link from the error report at the top of the page does not jump to the error.  This is a core bug and is not specific to this issue.
12. Add content to the "Body Text" field
13. Save the page with a revision log message
   - [x] Validate that it saves correctly.
14.  Edit the node.
15. **UNCheck** the box for "Create profile page with biography"
16. Save the page with a revision log message
   - [x] Validate that it saves correctly.

### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing announcement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping the UX writer so they are ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping the UX writer to write and queue content
- [ ] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

### CODE REVIEW

**Comment by swirtSJW** [comment]
Commands I used to restore the files from before #14233
```
git restore -s e4eb78311 -SW -- docroot/modules/custom/va_gov_backend/js/set_body_to_required.es6.js
git restore -s e4eb78311 -SW -- docroot/modules/custom/va_gov_backend/js/set_body_to_required.js
git restore -s e4eb78311 -SW -- docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php
git restore -s e4eb78311 -SW -- docroot/modules/custom/va_gov_backend/va_gov_backend.libraries.yml
git restore -s e4eb78311 -SW -- tests/behat/features/perms.feature
```

File: docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php [hunk 1]
```diff
@@ -61,8 +60,6 @@ public function __construct(
   public static function getSubscribedEvents(): array {
     return [
       'hook_event_dispatcher.form_node_centralized_content_edit_form.alter' => 'alterCentralizedContentNodeForm',
-      'hook_event_dispatcher.form_node_person_profile_edit_form.alter' => 'alterStaffProfileNodeForm',
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php:64]: This is all getting moved to the va_gov_profile module.

File: docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php [hunk 2]
```diff
@@ -125,67 +122,6 @@ private function trimNodeTitleWhitespace(NodeInterface $node) {
     $node->setTitle($title);
   }
 
-  /**
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php:128]: This is all getting moved to the va_gov_profile module.

File: config/sync/field.field.node.person_profile.field_body.yml [hunk 3]
```diff
@@ -8,16 +8,19 @@ dependencies:
...
+  tmgmt_content:
+    excluded: false
 id: node.person_profile.field_body
 field_name: field_body
 entity_type: node
 bundle: person_profile
 label: 'Body text'
 description: 'Add everything else from the second sentence on. Don''t include the first sentence in this field.'
-required: true
+required: false
```
Comment (swirtSJW) [inline config/sync/field.field.node.person_profile.field_body.yml:23]: Losing this here so we can handle it in Constraint elsewhere.

**Comment by swirtSJW** [comment]
Odd failure on behat.  Looking into it

```
Feature: Permissions

  @perms @content_editor
  Scenario Outline: Content editor can edit - edit test indicates save perm is true. # features/perms.feature:5
      """
      We may want to extend this test for other roles
      """
    Given I am logged in as a user with the <role> role                              # Drupal\DrupalExtension\Context\DrupalContext::assertAuthenticatedByRole()
    Then I should be able to edit an <type>                                          # Drupal\DrupalExtension\Context\DrupalContext::assertEditNodeOfType()

    Examples:
      | role             | type                               |
      | "content_editor" | "page"                             |
      | "content_editor" | "landing_page"                     |
      | "content_editor" | "health_care_region_detail_page"   |
      | "content_editor" | "event"                            |
      | "content_editor" | "event_listing"                    |
      | "content_editor" | "health_care_local_facility"       |
      | "content_editor" | "health_care_region_page"          |
      | "content_editor" | "health_care_local_health_service" |
      | "content_editor" | "press_release"                    |
      | "content_editor" | "office"                           |
      | "content_editor" | "outreach_asset"                   |
      | "content_editor" | "publication_listing"              |
      | "content_editor" | "person_profile"                   |
        Failed step: Then I should be able to edit an "person_profile"
        SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'title' cannot be null: INSERT INTO "node_field_data" ("nid", "vid", "type", "langcode", "status", "title", "uid", "created", "changed", "promote", "sticky", "default_langcode", "revision_translation_affected", "content_translation_source", "content_translation_outdated") VALUES (:db_insert_placeholder_0, :db_insert_placeholder_1, :db_insert_placeholder_2, :db_insert_placeholder_3, :db_insert_placeholder_4, :db_insert_placeholder_5, :db_insert_placeholder_6, :db_insert_placeholder_7, :db_insert_placeholder_8, :db_insert_placeholder_9, :db_insert_placeholder_10, :db_insert_placeholder_11, :db_insert_placeholder_12, :db_insert_placeholder_13, :db_insert_placeholder_14); Array
        (
            [:db_insert_placeholder_0] => 57846
            [:db_insert_placeholder_1] => 787338
            [:db_insert_placeholder_2] => person_profile
            [:db_insert_placeholder_3] => en
            [:db_insert_placeholder_4] => 0
            [:db_insert_placeholder_5] => 
            [:db_insert_placeholder_6] => 4621
            [:db_insert_placeholder_7] => 1688661931
            [:db_insert_placeholder_8] => 1688661931
            [:db_insert_placeholder_9] => 0
            [:db_insert_placeholder_10] => 0
            [:db_insert_placeholder_11] => 1
            [:db_insert_placeholder_12] => 
            [:db_insert_placeholder_13] => und
            [:db_insert_placeholder_14] => 0
        )
         (Drupal\Core\Entity\EntityStorageException)

```

**Comment by swirtSJW** [comment]
Code is working and locked in.  Re-working the angry tests right now.  It should not shift the code.

**Review by omahane — COMMENTED** [review]
I am not sure why the error is showing twice:
Screenshot 2023-07-06 at 1 16 54 PM

**Comment by swirtSJW** [comment]
> I am not sure why the error is showing twice: 

I am pretty sure that is because of re-writing of the field ids that CKeditor does when it takes over the value part of the fields.   However I can poke at that to test my assumption.

**Comment by swirtSJW** [comment]
Adding @ndouglas for eyes on the test addition and the new cypress step function for detecting html 5 errors.

**Comment by swirtSJW** [comment]
Odd failure on phpunit
```
There were 4 failures:

1) tests\phpunit\Content\TrimNodeTitleWhitespaceTest::testConfirmExpectedOutput with data set #1 ('This had trailing spaces    ', 'This had trailing spaces')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'This had trailing spaces'
+'This had trailing spaces    '
```

**Comment by swirtSJW** [comment]
Well at least cypress passed :)
image

**Comment by swirtSJW** [comment]
I figured out the odd failure.  It was a collision of eventsubscribers and the new one was wiping out the old one.  That was resolved with 
image

**Review by omahane — APPROVED** [review]
This is great! Works like a charm.

I'm blown away by what it took to get this back to working.

File: docroot/modules/custom/va_gov_profile/src/EventSubscriber/EntityEventSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+use Drupal\core_event_dispatcher\EntityHookEvents;
+use Drupal\core_event_dispatcher\Event\Entity\EntityTypeAlterEvent;
+use Drupal\core_event_dispatcher\Event\Form\FormIdAlterEvent;
+use Drupal\Core\Entity\EntityTypeManager;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\va_gov_user\Service\UserPermsService;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * VA.gov Generic Entity Event Subscriber. Do only multi-product stuff here.
```
Comment (ndouglas) [inline docroot/modules/custom/va_gov_profile/src/EventSubscriber/EntityEventSubscriber.php:14]: Is this comment accurate?

File: docroot/modules/custom/va_gov_profile/src/EventSubscriber/EntityEventSubscriber.php [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    $form = &$event->getForm();
+    $form['#attached']['library'][] = 'va_gov_profile/set_body_to_required';
+    $selector = ':input[name="field_complete_biography_create[value]"]';
+    $form['field_intro_text']['widget'][0]['value']['#states'] = [
+      'required' => [
+          [$selector => ['checked' => TRUE]],
+      ],
+      'visible' => [
+            [$selector => ['checked' => TRUE]],
+      ],
```
Comment (ndouglas) [inline docroot/modules/custom/va_gov_profile/src/EventSubscriber/EntityEventSubscriber.php:102]: Indenting here looks messed up. Surprised this isn't being caught by tests/hooks.

File: tests/cypress/integration/common/i_save_the_node.js [hunk 6]
```diff
@@ -7,3 +7,14 @@ Given("I save the node", () => {
...
+Given(
+  "I am prevented from saving the node by {string} {string} with error {string}",
+  (fieldType, selector, html5Error) => {
+    cy.get("form.node-form input#edit-submit").click({ force: true });
+    cy.get((fieldType += ":invalid")).should("have.length", 1);
+    cy.get(selector).then(($input) => {
+      expect($input[0].validationMessage).to.eq(html5Error);
+    });
+  }
+);
```
Comment (ndouglas) [inline tests/cypress/integration/common/i_save_the_node.js:20]: Nice!

**Comment by swirtSJW** [comment]
> I'm blown away by what it took to get this back to working.

Wellllll.... there was a bit easier way to get it back to working, but it was still as fragile as it had been and would have likely failed on occasion, so this was the way to make it more reliable in both function and data integrity.

**Comment by ndouglas** [comment]
>I am not sure why the error is showing twice:

That's [this](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/12415), right? 

I created a patch [here](https://www.drupal.org/project/drupal/issues/3229493) for that issue, but I think it's stopped working :weary:

**Review by ndouglas — DISMISSED** [review]
Beautiful.

File: docroot/modules/custom/va_gov_profile/src/Plugin/Validation/Constraint/PageRequiredFieldsConstraintValidator.php
Comment (ndouglas) [inline docroot/modules/custom/va_gov_profile/src/Plugin/Validation/Constraint/PageRequiredFieldsConstraintValidator.php:1]: I think this is a really good solution to the larger problem.

File: docroot/modules/custom/va_gov_profile/src/EventSubscriber/EntityEventSubscriber.php (hunk 4 above)
Comment (swirtSJW) [reply docroot/modules/custom/va_gov_profile/src/EventSubscriber/EntityEventSubscriber.php:14]: Bah.   No... soiled in copypasta.   :(

File: docroot/modules/custom/va_gov_profile/src/EventSubscriber/EntityEventSubscriber.php (hunk 5 above)
Comment (swirtSJW) [reply docroot/modules/custom/va_gov_profile/src/EventSubscriber/EntityEventSubscriber.php:102]: Good catch... tests aren't catching it nor is pre-commit hooks nor my ide.   Wierd.

File: docroot/modules/custom/va_gov_profile/src/Plugin/Validation/Constraint/PageRequiredFieldsConstraint.php [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+ * )
+ */
+class PageRequiredFieldsConstraint extends Constraint {
+
+  /**
+   * Message shown when validation fails.
+   *
+   * @var string
+   */
+  public $message = "In order to use '@checkbox' a profile page with biography, the @empty_fields must not be empty.";
```
[the remark this answers is not in the document]
Comment (swirtSJW) [reply docroot/modules/custom/va_gov_profile/src/Plugin/Validation/Constraint/PageRequiredFieldsConstraint.php:23]: Good call, the class name is very specific, and the only use for it is not.  I will adjust the class name.

**Comment by swirtSJW** [comment]
> > I am not sure why the error is showing twice:
> 
> That's [this](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/12415), right?
> 
> I created a patch [here](https://www.drupal.org/project/drupal/issues/3229493) for that issue, but I think it's stopped working weary

@ndouglas I think it is that or related to it, but your patch might be working on field validation, but failing since I am using entity validation.  Not sure, but a possibility.

**Comment by swirtSJW** [comment]
Thanks for the great review @ndouglas.  I addressed all your concerns in this [last commit](https://github.com/department-of-veterans-affairs/va.gov-cms/pull/14233/commits/f8d76b6f15fc2bacfaa0b4cf08cdf4ddc908075e) .

File: docroot/modules/custom/va_gov_profile/src/EventSubscriber/EntityEventSubscriber.php (hunk 4 above)
Comment (swirtSJW) [reply docroot/modules/custom/va_gov_profile/src/EventSubscriber/EntityEventSubscriber.php]: Fixed

File: docroot/modules/custom/va_gov_profile/src/EventSubscriber/EntityEventSubscriber.php (hunk 5 above)
Comment (swirtSJW) [reply docroot/modules/custom/va_gov_profile/src/EventSubscriber/EntityEventSubscriber.php]: fixed

File: docroot/modules/custom/va_gov_profile/src/Plugin/Validation/Constraint/PageRequiredFieldsConstraint.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (swirtSJW) [reply docroot/modules/custom/va_gov_profile/src/Plugin/Validation/Constraint/PageRequiredFieldsConstraint.php]: Fixed

**Review by ndouglas — APPROVED** [review]
❤️

File: config/sync/field.field.node.person_profile.field_body.yml (hunk 3 above)
Comment (edmund-dunn) [reply config/sync/field.field.node.person_profile.field_body.yml:23]: Like this idea!

**Review by edmund-dunn — APPROVED** [review]
Nicely done sir! I didn't thnk of this approach, but it is exactly what is needed! Thanks!

**Pull request merged**